hoomd/pytest/test_box.py

```diff
@@ -0,0 +1,205 @@
+        base_box.volume = v
+        assert np.allclose(aspect, get_aspect(base_box.L))
+        assert isclose(base_box.volume, v)
+
+
+def test_periodic(base_box):
+    assert all(base_box.periodic)
+
+
+@fixture
+def expected_matrix(box_dict):
+    return np.array([
+        [box_dict['Lx'],
+         box_dict['Ly'] * box_dict['xy'],
+         box_dict['Lz'] * box_dict['xz']],
+        [0, box_dict['Ly'], box_dict['Lz'] * box_dict['yz']],
+        [0, 0, box_dict['Lz']], ])
+
+
+def test_matrix(base_box, expected_matrix):
+    assert np.allclose(base_box.matrix, expected_matrix)
+    base_box.xy *= 2
+    assert isclose(base_box.matrix[0, 1], 2 * expected_matrix[0, 1])
+    base_box.yz *= 0.5
+    assert isclose(base_box.matrix[1, 2], 0.5 * expected_matrix[1, 2])
+    base_box.Lx *= 3
+    assert isclose(base_box.matrix[0, 0], 3 * expected_matrix[0, 0])
+
+
+@fixture
+def new_box_matrix_dict():
+    Lx, Ly, Lz = 2, 4, 8
+    xy, xz, yz = 1, 3, 5
+    new_box_matrix = np.array([
+        [Lx, Ly * xy, Lz * xz], [0, Ly, Lz * yz], [0, 0, Lz]])
+    return dict(Lx=Lx, Ly=Ly, Lz=Lz, xy=xy, xz=xz, yz=yz,
+                matrix=new_box_matrix)
+
+
+def test_matrix_setting(base_box, new_box_matrix_dict):
+    base_box.matrix = new_box_matrix_dict['matrix']
+    assert np.allclose(new_box_matrix_dict['matrix'], base_box.matrix)
+    assert np.allclose(base_box.L, [new_box_matrix_dict['Lx'],
+                                    new_box_matrix_dict['Ly'],
+                                    new_box_matrix_dict['Lz']])
+    assert np.allclose(base_box.tilts, [new_box_matrix_dict['xy'],
+                                        new_box_matrix_dict['xz'],
+                                        new_box_matrix_dict['yz']])
+
+
+def test_cube():
+    for L in np.linspace(1, 100, 10):
+        box = Box.cube(L)
+        assert all(box.L == L)
+        assert box.Lx == box.Ly == box.Lz == L
+
+
+def test_square():
+    for L in np.linspace(1, 100, 10):
+        box = Box.square(L)
+        assert all(box.L == [L, L, 0])
+        assert box.Lx == box.Ly == L and box.Lz == 0
+
+
+def test_from_matrix(new_box_matrix_dict):
+    box = Box.from_matrix(new_box_matrix_dict['matrix'])
+    assert np.allclose(new_box_matrix_dict['matrix'], box.matrix)
+    assert np.allclose(box.L, [new_box_matrix_dict['Lx'],
+                               new_box_matrix_dict['Ly'],
+                               new_box_matrix_dict['Lz']])
+    assert np.allclose(box.tilts, [new_box_matrix_dict['xy'],
+                                   new_box_matrix_dict['xz'],
+                                   new_box_matrix_dict['yz']])
+
+
+def test_eq(base_box, box_dict):
+    box2 = Box(**box_dict)
+    assert base_box == box2
+    box2.Lx = 2
+    assert not base_box == box2
+
+
+def test_neq(base_box, box_dict):
+    box2 = Box(**box_dict)
+    assert not base_box != box2
+    box2.Lx = 2
+    assert base_box != box2
+
+
+def test_copy(base_box):
+    copied_box = copy(base_box)
```

Comment:
`deepcopy` reports:
```
TypeError: cannot pickle 'hoomd._hoomd.BoxDim' object
```

Reply:
My main question here is whether the current behavior (which makes a new C++ object) is what is expected of `copy` vs `deepcopy`. I'm not sure we _need_ to implement these methods, but if we are going to I think we should put some thought into the expected behavior. For instance, I think I would expect `deepcopy` to do what `copy` is currently doing (make a new C++ box)`, while `copy` I'm not completely sure about.

Reply:
Anecdotally, in freud v1, boxes were pickle-able because we had the code snippet linked below. We took that ability to pickle boxes out in freud v2 for some reason (can't recall why). https://github.com/glotzerlab/freud/blob/e4787797a40ac3fb5ee9f11e76d430c37f6c7e62/freud/box.pyx#L636-L642

Reply:
The expected behavior depends on whether we think of a box as a "deep" or "shallow" object IMO. However, we can remove this and just use the `Box._from_cpp` if we want a quick copy, just a bit more verbose.